src/app/modules/examples/examples.service.ts

```diff
@@ -0,0 +1,327 @@
+import { get } from 'lodash'
+import mongoose from 'mongoose'
+import { errAsync, okAsync, ResultAsync } from 'neverthrow'
+
+import { createLoggerWithLabel } from '../../../config/logger'
+import getFormStatisticsTotalModel from '../../models/form_statistics_total.server.model'
+import getFormModel from '../../models/form.server.model'
+import getSubmissionModel from '../../models/submission.server.model'
+import { formatToRelativeString } from '../../utils/date'
+import { DatabaseError } from '../core/core.errors'
+
+import { MIN_SUB_COUNT, PAGE_SIZE } from './examples.constants'
+import { ResultsNotFoundError } from './examples.errors'
+import {
+  groupSubmissionsByFormId,
+  lookupFormStatisticsInfo,
+  lookupSubmissionInfo,
+  projectSubmissionInfo,
+  selectAndProjectCardInfo,
+} from './examples.queries'
+import {
+  ExamplesQueryParams,
+  FormInfo,
+  QueryData,
+  QueryExecResult,
+  QueryExecResultWithTotal,
+  QueryPageResult,
+  QueryPageResultWithTotal,
+  RetrievalType,
+  RetrieveSubmissionsExecResult,
+  SingleFormInfoQueryResult,
+  SingleFormResult,
+} from './examples.types'
+import {
+  createFormIdInfoPipeline,
+  createGeneralQueryPipeline,
+  createSearchQueryPipeline,
+  createSingleSearchStatsPipeline,
+  createSingleSearchSubmissionPipeline,
+} from './examples.utils'
+
+const FormModel = getFormModel(mongoose)
+const FormStatisticsModel = getFormStatisticsTotalModel(mongoose)
+const SubmissionModel = getSubmissionModel(mongoose)
+
+const logger = createLoggerWithLabel(module)
+
+/**
+ * Maps retrieval type to the middlewares and query model used for general
+ * queries to use when creating the aggregation pipeline
+ */
+const RETRIEVAL_TO_QUERY_DATA_MAP: QueryData = {
+  [RetrievalType.Stats]: {
+    generalQueryModel: FormStatisticsModel,
+    lookUpMiddleware: lookupFormStatisticsInfo,
+    groupByMiddleware: projectSubmissionInfo,
+    singleSearchPipeline: createSingleSearchStatsPipeline,
+  },
+  [RetrievalType.Submissions]: {
+    generalQueryModel: SubmissionModel,
+    lookUpMiddleware: lookupSubmissionInfo,
+    groupByMiddleware: groupSubmissionsByFormId,
+    singleSearchPipeline: createSingleSearchSubmissionPipeline,
+  },
+}
```

Comment:
this is when Dependency Injection would've been useful

Reply:
i agree. this was a pain and i still don't think this is the best implementation. Better than the old implementation imo though, haha!